src/pybind/mgr/orchestrator.py

```diff
@@ -0,0 +1,266 @@
+        assert isinstance(spec, StatelessServiceSpec)
+        raise NotImplementedError()
+
+    def update_stateless_service(self, service_type, id_, spec):
+        assert isinstance(spec, StatelessServiceSpec)
+        raise NotImplementedError()
+
+    def remove_stateless_service(self, service_type, id_):
+        raise NotImplementedError()
+
+    def upgrade_start(self, upgrade_spec):
+        assert isinstance(upgrade_spec, UpgradeSpec)
+        raise NotImplementedError()
+
+    def upgrade_status(self):
+        """
+        If an upgrade is currently underway, report on where
+        we are in the process, or if some error has occurred.
+
+        :return: UpgradeStatusSpec instance
+        """
+        raise NotImplementedError()
+
+    def upgrade_available(self):
+        """
+        Report on what versions are available to upgrade to
+
+        :return: List of strings
+        """
+        raise NotImplementedError()
+    # <<< from 21 March meeting
+
+    # >>> additional to 21mar: auto-placement of stateful services
+    def add_stateful_service_rule(self, service_type, stateful_service_spec,
+                                  placement_spec):
+        """
+        Stateful service rules serve two purposes:
+         - Optionally delegate device selection to the orchestrator
+         - Enable the orchestrator to auto-assimilate new hardware if it
+           matches the placement spec, without any further calls from ceph-mgr.
+
+        To create a confidence-inspiring UI workflow, use test_stateful_service_rule
+        beforehand to show the user where stateful services will be placed
+        if they proceed.
+        """
+        raise NotImplementedError()
+
+    def test_stateful_service_rule(self, service_type, stateful_service_spec,
+                                   placement_spec):
+        """
+        See add_stateful_service_rule.
+        """
+        raise NotImplementedError()
+
+    def remove_stateful_service_rule(self, service_type, id_):
+        """
+        This will remove the *rule* but not the services that were
+        created as a result.  Those should be converted into statically
+        placed services as if they had been created with add_stateful_service,
+        so that they can be removed with remove_stateless_service
+        if desired.
+        """
+        raise NotImplementedError()
+    # <<<
+
+
+class UpgradeSpec(object):
+    # Request to orchestrator to initiate an upgrade to a particular
+    # version of Ceph
+    def __init__(self):
+        self.target_version = None
+
+
+class UpgradeStatusSpec(object):
+    # Orchestrator's report on what's going on with any ongoing upgrade
+    def __init__(self):
+        self.in_progress = False  # Is an upgrade underway?
+        self.services_complete = []  # Which daemon types are fully updated?
+        self.message = ""  # Freeform description
+
+
+class PlacementSpec(object):
```

Comment:
This class could expose an abstract public method `def nodes(self)` that would return the list of nodes that this `PlacementSpec` object evaluates to. Then each subclass could implement different placement strategies, where the simplest one would be label based one. But I imagine that in a future we could have different placement strategies based on node/host attributes, for instance.

Reply:
It would need to be passed a list of `InventoryNode` to have anything to filter on, but yes.  However, the output would only be a range of possible locations: the actual location would be up to the orchestrator.

I'm imagining that in typical usage for e.g. MDS services, we'd just be passing a PlacementSpec that limits to nodes with a "ceph" label, so most of the time that list of possible locations would be quite large and not very meaningful.

Reply:
> I think being able to inspect what is about to happen is hugely important for a human user.

Agreed, insofar as it is possible.  We aren't going to be able to give the user a preview of where k8s will schedule their pods, so any UI on this needs to be prepared for the typical case where the preview location is "any of these 50 hosts", rather than expecting to be able to say specifically where something will be scheduled.

> another common one might be "nodes with label 'ceph' and with no ceph osds on it

I don't think this level of logic belongs here -- it's the orchestrator's job to schedule services, which would include being aware enough of resources to know which services can peacefully co-exist.

As with everything else, a user can force this stuff with some judicious label creation, but the principle is that in general they're deferring to the orchestrator on placement of stateless services.

Reply:
It is tricky. I was thinking about this all yesterday evening, and my thoughts have changed.

> Agreed, insofar as it is possible. We aren't going to be able to give the user a preview of where k8s will schedule their pods, so any UI on this needs to be prepared for the typical case where the preview location is "any of these 50 hosts", rather than expecting to be able to say specifically where something will be scheduled.

I also generally agree on this, at least the "insofar as it is possible". Imagining a CLI, I would be okay returning to the user "any of these 50 hosts: <list all>". For a UI, some amount of drill-down would be important. Like a field that says "This rule matches 50 hosts" with a "see more" link perhaps.

> I don't think this level of logic belongs here -- it's the orchestrator's job to schedule services, which would include being aware enough of resources to know which services can peacefully co-exist.

> As with everything else, a user can force this stuff with some judicious label creation, but the principle is that in general they're deferring to the orchestrator on placement of stateless services.

Agreed.

Reply:
I don't see a real value in providing a detailed list of all Ceph nodes in here. If it turns out that all implementations provide a `nodes` property, we can add it later on. Instead, I'd suggest to provide a `def __str__(self):` method that describes the location of this service.